crates/agent/src/discovers/specs.rs

```diff
@@ -0,0 +1,373 @@
+    let mut filtered_bindings = Vec::new();
+
+    for discovered_binding in discovered_bindings {
+        let DiscoveredBinding {
+            recommended_name,
+            resource_spec,
+            ..
+        } = &discovered_binding;
+
+        // Attempt to find a fetched binding such that the discovered resource
+        // spec is a strict subset of the fetched resource spec. In other
+        // words the fetched resource spec may have _extra_ locations,
+        // but all locations of the discovered resource spec must be equal.
+        let fetched_binding = fetched_bindings
+            .iter()
+            .filter(|fetched| {
+                doc::diff(
+                    Some(&serde_json::json!(&fetched.resource)),
+                    Some(&serde_json::json!(resource_spec)),
+                )
+                .is_empty()
+            })
+            .next();
+
+        if let Some(fetched_binding) = fetched_binding {
+            // Preserve the fetched version of a matched CaptureBinding.
+            capture_bindings.push(fetched_binding.clone());
+            filtered_bindings.push(discovered_binding);
+        } else if !update_only {
+            // Create a new CaptureBinding.
+            capture_bindings.push(models::CaptureBinding {
+                target: models::Collection::new(format!("{capture_prefix}/{recommended_name}")),
+                resource: resource_spec.clone(),
+            });
+            filtered_bindings.push(discovered_binding);
+        }
+    }
+
+    (
+        models::CaptureDef {
+            endpoint,
+            bindings: capture_bindings,
+            interval,
+            shards,
+        },
+        filtered_bindings,
+    )
+}
+
+pub fn merge_collections(
+    discovered_bindings: Vec<DiscoveredBinding>,
+    mut fetched_collections: BTreeMap<models::Collection, models::CollectionDef>,
+    targets: Vec<models::Collection>,
+) -> BTreeMap<models::Collection, models::CollectionDef> {
+    assert_eq!(targets.len(), discovered_bindings.len());
+
+    let mut collections = BTreeMap::new();
+
+    for (
+        target,
+        DiscoveredBinding {
+            key_ptrs,
+            document_schema,
+            ..
+        },
+    ) in targets.into_iter().zip(discovered_bindings.into_iter())
+    {
+        // Unwrap a fetched collection, or initialize a blank one.
+        let mut collection =
+            fetched_collections
+                .remove(&target)
+                .unwrap_or_else(|| models::CollectionDef {
+                    schema: None,
+                    write_schema: None,
+                    read_schema: None,
+                    key: models::CompositeKey::new(Vec::new()),
+                    projections: Default::default(),
+                    derivation: None,
+                    journals: Default::default(),
+                });
+
+        if collection.read_schema.is_some() {
+            collection.write_schema = Some(document_schema);
```

Comment:
What if the fetched collection already has a write schema? It looks like this will overwrite it? 

Or maybe the thinking is that there's never a circumstance where the write schema would be different from the document schema... but that can't be true because the whole point of write schemas was to allow for more permissive writes and resolve potential schema issues asynchronously on the read side.

Reply:
It will be over-written by this, correct.

Conceptually, the capture connector "owns" the write schema. Connectors SHOULD implement fixed write schemas which are known to always-work (like allowing any json document, so long as it has meta fields of the filename and offset in, say, an S3 connector).

The read schema is owned by the user, and updated manually or perhaps through inference of actually-captured documents. That we definitely _don't_ want to over-write with the connector output, which is what happens today.